backends/platform/atari/readme.txt

```diff
@@ -0,0 +1,422 @@
+operation:
+
+1. "True triple buffering" as described in
+https://en.wikipedia.org/wiki/Multiple_buffering#Triple_buffering (vsync on)
+
+2. "Swap chain" as described in https://en.wikipedia.org/wiki/Swap_chain (vsync
+off)
+
+Pros:
+
+- best compromise between performance and visual experience
+
+- works well with both higher and lower frame rates
+
+Cons:
+
+- since three buffers are present, the buffer is always blitted into the screen
+  surface as whole, even if only one tiny little rectangle is changed (excluding
+  the cursor)
+
+- slightly irregular frame rate (depends solely on the game's complexity)
+
+- in case of extremely fast rendering in 1.), one or more buffers are
+  dropped in favor of showing only the most recent one (unlikely)
+
+- in case of extremely fast rendering in 2.), screen tearing is possible
+  because the rendering pipeline starts overwriting the buffer which is
+  currently displayed (unlikely)
+
+SuperBlitter used: yes, for rectangle blitting to screen and cursor restoration.
+Sometimes also for generic copying between buffers (see above).
+
+Triple buffering with vsync on is the default mode for this port.
+
+
+SuperVidel and SuperBlitter
+---------------------------
+
+As mentioned, this port uses SuperVidel and its SuperBlitter heavily. That
+means that if the SuperVidel is detected, it does the following:
+
+- patches all 8bpp VGA resolutions to chunky ones, rendering all C2P routines
+  useless
+
+- patches all surface addresses with OR'ing 0xA0000000, i.e. using SV RAM
+  instead of slow ST RAM (and even instead of TT RAM for allowing pure
+  SuperBlitter copying)
+
+- when SuperVidel FW version >= 9 is detected, the async FIFO buffer is used
+  instead of the slower sync blitting (where one has to wait for every
+  rectangle blit to finish). This applies only for chunky buffer -> screen
+  surfaces copy (as the generic surface copy can't rely on this behavior) but
+  despite this limitation it sometimes leads to nearly zero-cost rendering
+  and makes a *huge* difference for 640x480 fullscreen updates.
+
+
+Performance considerations/pitfalls
+-----------------------------------
+
+It's important to understand what affects performance on our limited platform
+to avoid unpleasant playing experiences.
+
+Game engines with unexpected performance hit
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A typical example from this category is the Gobliins engine (and its
+sequels). At first it looks like our machine / backend is doing something
+terribly wrong but the truth is it is the engine itself which is doing a lot of
+unnecessary redraws and updates, sometimes even before reaching the backend.
+The only real solution is to profile and fix the engine.
+
+Too many fullscreen updates
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Somewhat related to the previous point - sometimes the engine authors didn't
+realize the impact of every update on the overall performance and instead of
```

Comment:
It is up to you to skip `updateScreen()` calls. The engine authors are encouraged to call it often.

You may implement frame rate locking, e.g. skip blotting, if not enough time has passed. The flow of the `updateScreen()` calls is constant, even when games are paused (we have to do it to make mouse redraws).

Reply:
It's not like the calling itself is the problem, what I meant is that sometimes I saw that a game updated one rectangle and called updateScreen(), immediately after that updated another rectangle and called updateScreen() yet again.

But the idea of actually ignoring (queuing) the updates according to frame rate is brilliant! I'll add it to my TODO list.